Capstone Full Software Stack Rev0

SamKhzym opened the pull request:
initial acc controller functionality implemented by @petruntn.

petruntn committed:
initial acc commit

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+#include<stdio.h>
+
+//Globals
+float hostVel;
+float hostVelPrev; //Host velocity in the previous time step
+float integral = 0; //Needs to be global as it is used in the calculation of next step
+float errorPrev = -999.0; //Previous value of PID error
+float leadDistPrev = 0;
+float overshoot = 0;
+float leadTime = 0;
```
lots of global variables to be hanging around... why not combine into a struct for the state of the system? would be nice to have different structs for different parameter sets (ie. PID params, calibrations, etc.)

ie.
```C
struct pidParams {
  float integralError;
  char prevError;
  ...
};

struct accParams {
  pidParams speedPid;
  ...
}
```

PradCoder replied:
Not sure  exactly what attributes the structs should have (what exactly should pid have and what should acc have)

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+float leadTime = 0;
+
+//Calibratables
+float timeStep = 0.1; //Time step of system
+float simTime = 20.0; //Simulation time in seconds
```
may move these to a different header file... maybe `calibrations.h` or smth...

PradCoder replied:
Made a header file with function headers, moved enums, structs, and consts

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+float timeStep = 0.1; //Time step of system
+float simTime = 20.0; //Simulation time in seconds
+
+void vehDynamics(float torque){
+    //torque will be a value from -1 to 1, where negative values represent braking torque and positive values represent acceleration
+
+    //Calculate drag
+    //CdA = 0.27
+    float CdA = 0.55; //Default value for Drag Area
+    float rho = 1.293; //Default for air density
+    float mass = 1670.0; //Default of mass of vehicle
+
+    float drag = -0.5*rho*CdA*hostVel*hostVel/mass; //Deceleration due to drag
+
+    float hostAccel;
+    if (torque >= 0){ //Braking is more powerful than acceleration. Will this break the PID? Probably
+        hostAccel = drag + torque*4; //Assuming max acceleration is 4m/s; may want to make this a lookup table at some point
+    }
+    else if(torque < 0){
+        hostAccel = drag + torque*7; //Assuming max deceleration is 7m/s; may want to make this a lookup table at some point
+    }
+    
+    hostVel += hostAccel*timeStep; //Iterate the host velocity based on acceleration
+
+    //Prevent negative velocity
+    if (hostVel < 0){
+        hostVel = 0;
+    }
+}
```
nice model! maybe isolate to its own file... we don't necessarily want to compile in the vehicle dynamics into the core controller. that's going to be implemented in the simulator.

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+
+float accAlgo(float hostVelInit, float leadVel, float setSpeed, float leadDist){
+    //Decide between cruise and follow modes
+    int mode; //0 = cruise, 1 = follow
```
enum?

PradCoder replied:
Added enums for this, but I need to organize the code a bit more

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+float accAlgo(float hostVelInit, float leadVel, float setSpeed, float leadDist){
+    //Decide between cruise and follow modes
+    int mode; //0 = cruise, 1 = follow
+    if (leadVel > setSpeed || leadDist > 500){ //Future additions: safety critical mode, failure mode
```
where does 500 come from? arbitrary? that's pretty far for a lead vehicle, generally max sensing range is ~200m...

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+        setSpeed = leadVel;
+    }
+    
+    //Calculate PID terms
+    float error = setSpeed - hostVel; //P
+    integral += error*timeStep; //I
+    float derivative = (error - errorPrev)/timeStep; //D
+
+    //Set PID parameters
+    float kP = 3;
+    float kI = 0.05;
+    float kD = 0.04;
```
can this be moved to its own function? would be nice to reuse pid functionality

ie.
```C
float pidStep(float error, pidParams* params) {
  ...
  return command;
}
```

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+    float kI = 0.05;
+    float kD = 0.04;
+
+    //Remove derivative from first time step
+    if (errorPrev == -999.0){ //AKA first time step
+        derivative = 0;
+    }
```
icky... can be resolved by separating acc function into two functions -- an `accInit()` which runs once initially and sets all the globals and `accStep()` which runs periodically.

PradCoder replied:
Made an init, but it takes errorPrev, and derivative as arguments (I still need to separate the globals variables)

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+    //Implement PID controller
+    float torque = kP * error + kI * integral + kD * derivative;
+
+    //Saturate torque value
+    if (torque >= 1.00){
+        torque = 1.0;
+    }
+    else if (torque <= -1.00){
+        torque = -1.0;
+    }
```
may want to move max torque outputs to calibrations file... also look into PID anti-winding for making sure error doesn't accumulate like crazy in the integral if the output is being saturated.

http://brettbeauregard.com/blog/2011/04/improving-the-beginners-pid-reset-windup/

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+    dynamics[0] = leadVel;
+    dynamics[1] = leadDist;
+}
+int main (){
+    //Define Calibrations
+    float setSpeed = 50/3.6;
+    float hostVelInit = 50/3.6;
+    float leadVelInit = 50/3.6;
+    float leadDistInit = 1000;
+    int mode = 1; //0 = Constant, 1 = Linear, 2 = Stepwise, 3 = Parabolic, 4 = Sinusoidal
+    hostVel = hostVelInit;
+
+    printf("Time| Error   | Torque  | HostVel | LeadDist| LeadVel|Mode\n");
+    int steps = simTime/timeStep;
+    for (int i = 0; i <= steps; i++){
+        printf("%04.1f| ", i*timeStep);
+
+        //Calculate the lead vehicle speed and distance
+        float leadDyn[2];
+        leadDynamics(leadDyn, i*timeStep, leadDistInit, leadVelInit, mode);
+
+        //Initial Speed, Lead Velocity, Set Speed, Lead Distance
+        accAlgo(hostVelInit, leadDyn[0], setSpeed, leadDyn[1]);
+        
+    }
+
+    printf("\nMax overshoot: %f\nLead Time: %f",-1*overshoot, leadTime);
+    return 0;
+}
```
may move main test file to its own file as well... don't want to compile this main for the target hardware since we don't want lead dynamics running there

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+    return torque;
+}
+
+float leadDynamics(float* dynamics, float timeStamp, float leadDistInit, float leadVelInit, int mode){
```
same comment as with vehDynamics...

SamKhzym commented on ACC Algorithm/ACCAlgorithMain.c:
```diff
@@ -0,0 +1,154 @@
+    errorPrev = error;
+
+    //Implement PID controller
+    float torque = kP * error + kI * integral + kD * derivative;
```
no distance control, just speed? what happens if lead vehicle is 5m away but we're going at same speed? this controller will maintain speed even at the risk of being way too close.

SamKhzym reviewed (commented):
good start, could definitely benefit from some separation into different files (core logic vs. helpers/dynamics models)... also could benefit from making some structs for code organization. make sure you implement the struct definitions in a separate header file so other code can include them as well! helps for logging. also some general controller comments, distance control def needs to be included in some form.

PradCoder committed:
Create build.yml

PradCoder committed:
Update build.yml

PradCoder committed:
Changed the names of somethings

PradCoder committed:
Added enums and some structs to begin working on

PradCoder committed:
Refactored to use header file, left global variables

PradCoder committed:
Added comments for ACC

SamKhzym committed:
removed software unnecessary docs

SamKhzym committed:
restrucutred acc code

SammerTheHammer77 committed:
initial road viz

SammerTheHammer77 committed:
modified viz to be in the perspective of the ego vehicle

SammerTheHammer77 committed:
viz works kinda with simple rectangle for lead car and moving lane lines to indicate ego speed

SammerTheHammer77 committed:
added stat display to viz window

SammerTheHammer77 committed:
added speedometer code for arduinio (untested)

SammerTheHammer77 committed:
removed viz folder, unlikely we will have any more viz

SammerTheHammer77 committed:
removed h file from cpp

SammerTheHammer77 committed:
cleaned up and added functionality to speedometer class and main loop

SammerTheHammer77 committed:
fixed viz bug with distance to distortion

SammerTheHammer77 committed:
data processing script for arduino serial monitor vals

SammerTheHammer77 committed:
wrote schema functions for serializing/deserializing compound data in C/C++. updated arduino code to fix bugs and improve main loop. arduino code does not build, needs debugging....

petruntn committed:
Modified ACC code to reflect actuation request instead of torque

PradCoder committed:
introduced test suite + refactor

PradCoder committed:
merged test suite submodule
Merge branch 'acc_initial' of https://github.com/SamKhzym/capstone into acc_initial

SammerTheHammer77 committed:
Merge branch 'acc_initial' of github.com:SamKhzym/capstone into acc_initial

SammerTheHammer77 committed:
merge onflict

SammerTheHammer77 committed:
ACC restructure + cleanup

SammerTheHammer77 committed:
Merge branch 'viz' into acc_initial

SammerTheHammer77 committed:
wrote the lead simulator and updated viz interface

SammerTheHammer77 committed:
separated simulator into sil and hil entry points, created plant model (2nd order tf), added requirements.txt and stuff to .gitignore

SammerTheHammer77 committed:
temp changes to allow arduino code to compile for speed testing

SammerTheHammer77 committed:
created wrapper for acc dll in python. full sil setup works!!

SammerTheHammer77 committed:
improving robustness of arduino speedometers, wrote analysis scripts for empirical tf derivation. don't have a good representative model yet
